notify/notify.go

```diff
@@ -119,109 +143,271 @@ func Now(ctx context.Context) (time.Time, bool) {
 	return v, ok
 }
 
-// A Notifier is a type which notifies about alerts under constraints of the
-// given context.
-type Notifier interface {
-	Notify(context.Context, ...*types.Alert) error
+// A Stage processes alerts constraints under the constraints of the given
+// context.
+type Stage interface {
+	Exec(ctx context.Context, alerts ...*types.Alert) ([]*types.Alert, error)
+}
+
+// Builds a RoutingStage as a map of receivers to Stages
+func BuildRoutingStage(confs []*config.Receiver, tmpl *template.Template, meshWait func() time.Duration, inhibitor *inhibit.Inhibitor, silences *meshprov.Silences, ni *meshprov.NotificationInfos, marker types.Marker) RoutingStage {
```

Comment:
I would argue that the name is a bit misleading here. In fact this function calls `createStage` and hence builds the entire nesting of stages. 

You call the result `pipeline` in `main()`. So maybe `BuildPipeline` might actually be a fitting name.